vespalib/src/vespa/vespalib/data/smart_buffer.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "smart_buffer.h"
+#include <cassert>
+
+namespace vespalib {
+
+void
+SmartBuffer::ensure_free(size_t bytes)
+{
+    if (write_len() >= bytes) {
+        return;
+    }
+    if ((unused() < bytes) || ((unused() * 3) < read_len())) {
+        size_t new_size = std::max(_data.size() * 2, read_len() + bytes);
+        alloc::Alloc new_buf(alloc::Alloc::alloc(new_size));
+        memcpy(new_buf.get(), read_ptr(), read_len());
+        _data.swap(new_buf);
+    } else {
+        memmove(_data.get(), read_ptr(), read_len());
+    }
+    _write_pos = read_len();
+    _read_pos = 0;
+}
+
+SmartBuffer::SmartBuffer(size_t initial_size)
+    : _data(alloc::Alloc::alloc(initial_size)),
+      _read_pos(0),
+      _write_pos(0)
+{
+}
+
+SmartBuffer::~SmartBuffer() = default;
+
+Memory
+SmartBuffer::obtain()
+{
+    return Memory(read_ptr(), read_len());
+}
+
+Input &
+SmartBuffer::evict(size_t bytes)
+{
+    assert(read_len() >= bytes);
+    _read_pos += bytes;
+    if (_read_pos == _write_pos) {
+        _read_pos = 0;
+        _write_pos = 0;
```

Comment:
Are overlapped obtains/reserves prohibited, or is that expected to work? This seems to imply the former, as it could reset the write cursor when there's a reserved area active. If so, maybe it should be added as an explicit comment to the class?

Reply:
I also had this thought when I wrote the code, but realized that the fact that we keep the data in a single continuous memory region already disallows overlapping input/output operations since calling reserve may need to relocate the data into a bigger buffer. The API is defined by the combination of the Input and Output interfaces. As you said; you can be both, but not at the same time (interleaved). Kind of like iterator invalidation in container classes. Maybe we should add documentation directly to the Input/Output interfaces instead...